enhancements/node-pull-credentials/pull-credentials.md

```diff
@@ -0,0 +1,252 @@
+
+### Implementation Details/Notes/Constraints
+
+This is the node's filesystem path we are taking into account:
+
+- `/var/lib/kubelet/config.json` contains the node's pull secret.
+
+#### Image Stream Import
+
+- Mount the pull secret as `readOnly`, `hostPath` in
+  `openshift-apiserver` deployment under `/var/lib/kubelet/config.json`.
+
+```yaml
+volumeMounts:
+- mountPath: /var/lib/kubelet/config.json
+  name: node-pull-credentials
+  readOnly: true
+volumes:
+- name: node-pull-credentials
+  hostPath:
+    path: /var/lib/kubelet/config.json
+    type: File
+```
+
+- When a user imports an imagestream, `openshift-apiserver` will merge all pull
+  credentials found in `/var/lib/kubelet/config.json` with other credentials
+  that may exist in the namespace.
+
+
+#### Builds
+
+- `controller-manager` would mount pull credentials inside `build` pod as 
+  `hostPath`(similar to what is done for Image Stream Import).
+- Builder image parses node's pull credentials and uses them during build,
+  merging with other pull credentials that are linked to the `builder` service
+  account. If the `BuildConfig` specifies a pull secret, we will continue the
+  current behavior of using the provided pull secret as an override.
+
+#### Registry pull-through
+
+- As done for Image Stream Import, mount pull credentials inside the image
+  registry pod.
+- Pull credentials will then be consumed by the registry.
+
+### Risks and Mitigations
+
+#### Pull credentials for registry may exist on namespace and on node
+
+User may have created a secret containing credentials for a registry that node
+already has credentials to.
+
+Mitigations:
+
+Always prioritize namespace secrets over node's credentials. If a credential
+exists on the namespace, do not use node's credentials.
+
+#### Pull credentials inside build pod may be visible to the user
+
+Cluster wide credentials mounted inside the builder pod may be a security risk
+as the user may be able to shell into the pod and copy them. Another possible
+risk is that the user may copy the credentials into a resulting image.
+
+Mitigations:
+
+As far as I verified(and this needs to be once more tested) it is impossible
+to spawn a shell inside the builder pods. I also tried to copy the credentials
+from builder's filesystem into a resulting image but it failed.
+
+#### Pushing to registries using node's credentials
+
+If pull credentials are used during builds we may allow users to push images to
+any registry configured on the node. This is not an ideal scenario as cluster
+administrators may not to be aware or wanting this.
+
+Mitigations
+
+Pull credentials must be used **only** for pulling images. On pushing we must
+use only user provided(not node) credentials'.
+
+#### Absence of mounted path on node filesystem
+
+If the path we are trying to mount through `hostPath` directive does not exist
+on the node where the pod is running the pod won't come up.
+
+Mitigations
+
+Mounted path always exist on worker nodes.
```

Comment:
what about master nodes?

Reply:
Interesting. The confirmation we have is that this path does not exist [only on bootstrap nodes](https://github.com/openshift/enhancements/pull/136#discussion_r368620589). @adambkaplan could we tag someone from node's team to get a final answer on this?

Reply:
@cgwalters can you please confirm that the cluster pull secret always exists at `/var/lib/kubelet/config.json`

Reply:
I don't want to make that permanent "ABI" though; see also https://github.com/openshift/machine-config-operator/issues/1190 which proposes ensuring all MCO configuration (including pull secret) lives in `/etc` so it can be made transactional - (but we'd probably have a backcompat symlink for a while at least).

I think probably the best bet is for whatever wants it to parse the kubelet config or so perhaps?

Or, I guess we could make `/etc/kubernetes/pull-secret.json -> /var/lib/kubelet/config.json` today and people could start using the new name.  (Or probably better introduce the that new `/etc` name upstream)

Reply:
> I don't want to make that permanent "ABI" though; see also [openshift/machine-config-operator#1190](https://github.com/openshift/machine-config-operator/issues/1190) which proposes ensuring all MCO configuration (including pull secret) lives in `/etc` so it can be made transactional - (but we'd probably have a backcompat symlink for a while at least).

It is clear that we must make the file path easily configurable.
 
> I think probably the best bet is for whatever wants it to parse the kubelet config or so perhaps?

@cgwalters Do you mean to parse `/etc/kubernetes/kubelet.conf` to know from where to read pull secrets? I could not find any directive on it pointing to `/var/lib/kubelet/config.json`, pardon if I misunderstood you, could you clarify?

> Or, I guess we could make `/etc/kubernetes/pull-secret.json -> /var/lib/kubelet/config.json` today and people could start using the new name. (Or probably better introduce the that new `/etc` name upstream)

I guess this would take some time to get done and we already have some PRs in place to implement this enhancement. I am inclined to stick with `/var/lib/kubelet/config.json`, making it easily configurable so we can easily migrate in the future.

Reply:
@cgwalters are you going to change root-dir for kubelet?